Fix flaky test TestClusterStatusMonitorLifecycle

LOG:
> 2020-08-14T23:50:43.0368859Z [ERROR]   TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle:293 expected:<true> but was:<false>

>2020-08-14T23:50:42.7829002Z [ERROR] testClusterStatusMonitorLifecycle(org.apache.helix.monitoring.TestClusterStatusMonitorLifecycle)  Time elapsed: 6.505 s  <<< FAILURE!
2020-08-14T23:50:42.7829267Z java.lang.AssertionError: expected:<true> but was:<false>
2020-08-14T23:50:42.7829704Z 	at org.apache.helix.monitoring.TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle(TestClusterStatusMonitorLifecycle.java:293)

code:
Should use TestHelper.DEFAULT_TIMEOUT, not it 60sec. 
Examine all TestHelper.verify(), there is no downside to use a more conservative value, or not?

```
    Assert.assertTrue(TestHelper.verify(() -> ManagementFactory.getPlatformMBeanServer()
        .queryMBeans(new ObjectName("ClusterStatus:*"), exp2).isEmpty(), 3000));
  }

```

LOG

>2020-09-25T05:34:14.4061345Z [ERROR] testClusterStatusMonitorLifecycle(org.apache.helix.monitoring.TestClusterStatusMonitorLifecycle)  Time elapsed: 64.82 s  <<< FAILURE!
2020-09-25T05:34:14.4067624Z java.lang.AssertionError: expected:<true> but was:<false>
2020-09-25T05:34:14.4080734Z 	at org.apache.helix.monitoring.TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle(TestClusterStatusMonitorLifecycle.java:298)
2020-09-25T05:34:14.4084644Z 
2020-09-25T05:34:14.8023392Z [ERROR] Failures: 
2020-09-25T05:34:14.8025868Z [ERROR]   TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle:298 expected:<true> but was:<false>
2020-09-25T05:34:14.8028368Z [ERROR] Tests run: 1205, Failures: 1, Errors: 0, Skipped: 1

--- a/helix-core/src/test/java/org/apache/helix/manager/zk/client/TestHelixZkClient.java
+++ b/helix-core/src/test/java/org/apache/helix/manager/zk/client/TestHelixZkClient.java
@@ -94,11 +94,12 @@ public class TestHelixZkClient extends ZkUnitTestBase {
     HelixZkClient sharedZkClientA =
         testFactory.buildZkClient(connectionConfig, new HelixZkClient.ZkClientConfig());
     Assert.assertTrue(sharedZkClientA.waitUntilConnected(1, TimeUnit.SECONDS));
+    long sessionIdClientA = sharedZkClientA.getSessionId();
 
     HelixZkClient sharedZkClientB =
         testFactory.buildZkClient(connectionConfig, new HelixZkClient.ZkClientConfig());
     Assert.assertTrue(sharedZkClientB.waitUntilConnected(1, TimeUnit.SECONDS));
-
+    long sessionIdClientB = sharedZkClientB.getSessionId();
     Assert.assertEquals(testFactory.getActiveConnectionCount(), 1);
 
     // client A and B is sharing the same session.
@@ -131,12 +132,16 @@ public class TestHelixZkClient extends ZkUnitTestBase {
     });
 
     // Modify using client A and client B will get notification.
+    // note, if session changed in between, we may get a dataDelete callback. This is current
+    // zkclient behavior.
     sharedZkClientA.createPersistent(TEST_PATH, true);
-    Assert.assertTrue(TestHelper.verify(() -> notificationCountB[0] == 1, 1000));
+    Assert.assertTrue(TestHelper.verify(() -> notificationCountB[0] == 1, TestHelper.WAIT_DURATION),
+        String.format("Original sidA before %d and sidB %d, current sidA %d, sidB %d",
+            sessionIdClientA, sessionIdClientB, sharedZkClientA.getSessionId(), sharedZkClientB.getSessionId()));
     Assert.assertEquals(notificationCountB[1], 0);
 
     sharedZkClientA.deleteRecursively(TEST_ROOT);
-    Assert.assertTrue(TestHelper.verify(() -> notificationCountB[1] == 1, 1000));
+    Assert.assertTrue(TestHelper.verify(() -> notificationCountB[1] == 1, TestHelper.WAIT_DURATION));
     Assert.assertEquals(notificationCountB[0], 1);
 
     try {
@@ -166,8 +171,8 @@ public class TestHelixZkClient extends ZkUnitTestBase {
     sharedZkClientB.watchForData(TEST_PATH);
     // Now modify using client B, and client A won't get notification.
     sharedZkClientB.createPersistent(TEST_PATH, true);
-    Assert.assertTrue(TestHelper.verify(() -> notificationCountB[0] == 2, 1000));
-    Assert.assertFalse(TestHelper.verify(() -> notificationCountA[0] == 2, 1000));
+    Assert.assertTrue(TestHelper.verify(() -> notificationCountB[0] == 2, TestHelper.WAIT_DURATION));
+    Assert.assertFalse(TestHelper.verify(() -> notificationCountA[0] == 2, TestHelper.WAIT_DURATION));
     sharedZkClientB.deleteRecursively(TEST_ROOT);
 
     Assert.assertEquals(testFactory.getActiveConnectionCount(), 1);

--- a/helix-core/src/test/java/org/apache/helix/monitoring/TestClusterStatusMonitorLifecycle.java
+++ b/helix-core/src/test/java/org/apache/helix/monitoring/TestClusterStatusMonitorLifecycle.java
@@ -217,6 +217,7 @@ public class TestClusterStatusMonitorLifecycle extends ZkTestBase {
 
     ZkHelixClusterVerifier controllerClusterVerifier =
         new BestPossibleExternalViewVerifier.Builder(_controllerClusterName).setZkClient(_gZkClient)
+            .setWaitTillVerify(TestHelper.DEFAULT_REBALANCE_PROCESSING_WAIT_TIME)
             .build();
     Assert.assertTrue(controllerClusterVerifier.verifyByPolling(),
         "Controller cluster was not converged");
@@ -291,7 +292,12 @@ public class TestClusterStatusMonitorLifecycle extends ZkTestBase {
         Query.not(Query.match(Query.attr("SensorName"), Query.value("MessageQueueStatus.*"))),
         exp1);
 
-    Assert.assertTrue(TestHelper.verify(() -> ManagementFactory.getPlatformMBeanServer()
-        .queryMBeans(new ObjectName("ClusterStatus:*"), exp2).isEmpty(), 3000));
+    // Note, the _asyncTasksThreadPool shutting down logic in GenericHelixController is best effort
+    // there is not guarantee that all threads in the pool is gone. Mossstly they will, but not always.
+    // see https://github.com/apache/helix/issues/1280
+    boolean result = TestHelper.verify(() -> ManagementFactory.getPlatformMBeanServer()
+        .queryMBeans(new ObjectName("ClusterStatus:*"), exp2).isEmpty(), TestHelper.WAIT_DURATION);
+    Assert.assertTrue(result,
+        "A small chance this may fail due to _asyncThread pool in controller may not shutdown in time. Please check issue 1280 to verify if this is the case.");
   }
 }

